### PR TITLE
[feature]SparkLoad support lake table

### PR DESCRIPTION
**Pull request by abc982627271**
## What type of PR is this：
- [ ] bug
- [x] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] I have added user document for my new feature or new function

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java [hunk 1]
```diff
@@ -475,70 +475,72 @@ private Set<Long> submitPushTasks() throws UserException {
...
-                                            tBrokerScanRange.getBroker_addresses().add(
-                                                new TNetworkAddress(fsBroker.ip, fsBroker.port));
-                                            LOG.debug("push task for replica {}, broker {}:{}, backendId {}," + 
-                                                    "filePath {}, fileSize {}", replicaId, fsBroker.ip, fsBroker.port, 
-                                                    backendId, tBrokerRangeDesc.path, tBrokerRangeDesc.file_size);
-                                        } else {
-                                            LOG.debug("push task for replica {}, backendId {}, filePath {}, fileSize {}",
-                                                    replicaId, backendId, tBrokerRangeDesc.path,
-                                                    tBrokerRangeDesc.file_size);
+                                PushBrokerReaderParams params = getPushBrokerReaderParams(table, indexId);
```
Comment (chaoyli) [inline fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java:481]: What's the purpose of the pr? Feature/BugFix/Enhancement or Refactor. You have to describe it in the commit message.

**Review by chaoyli — DISMISSED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 41 / 78 (52.56%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[466, 498, 499, 534, 535, 536, 537, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|40|57|70.18%|[502, 503, 504, 505, 506, 509, 510, 511, 512, 513, 514, 516, 517, 518, 603, 604, 605]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 41 / 78 (52.56%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[466, 498, 499, 534, 535, 536, 537, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|40|57|70.18%|[502, 503, 504, 505, 506, 509, 510, 511, 512, 513, 514, 516, 517, 518, 603, 604, 605]|

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java [hunk 2]
```diff
@@ -461,6 +463,7 @@ private void finishRealtimePush(AgentTask task, TFinishTaskRequest request) {
                     Replica replica =
                             findRelatedReplica(olapTable, partition, backendId, tabletId, tabletMeta.getIndexId());
                     // if the replica is under schema change, could not find the replica with aim schema hash
+                    LOG.info("replica: {}", replica);
```
Comment (tracymacding) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:466]: remove it

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java [hunk 3]
```diff
@@ -521,17 +530,28 @@ private Replica findRelatedReplica(OlapTable olapTable, Partition partition,
...
-            LOG.warn("could not find replica with backend {} in tablet {} in rollup index {} ",
-                    backendId, tabletId, indexId);
+
+        Tablet tablet = index.getTablet(tabletId);
+        if (tablet instanceof LakeTablet) {
+            List<Replica> allQueryableReplicas = Lists.newArrayList();
+            tablet.getQueryableReplicas(allQueryableReplicas, Collections.emptyList(),
+                    0, -1, 0);
+            if (allQueryableReplicas.isEmpty()) {
+                LOG.warn("no queryable replica found in tablet {}", tabletId);
```
Comment (tracymacding) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:540]: should return null

File: be/src/storage/lake/spark_load.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    Status st = Status::OK();
+    if (push_type == PUSH_FOR_DELETE) {
+        return Status::NotSupported("does not support push delete for lake table");
+    }
+
+    if (push_type == PUSH_NORMAL_V2) {
+        st = _load_convert(tablet, request);
+    }
+
+    return Status::OK();
```
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:56]: if (push_type == PUSH_NORMAL_V2) {
    return _load_convert(tablet, request);
} else {
    return Status::NotSupported(fmt::format("unsupported push type for lake table: {}", push_type));
}

File: be/src/storage/lake/spark_load.cpp [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+}
+
+Status SparkLoadHandler::_load_convert(Tablet& cur_tablet, const TPushReq& request) {
+    Status st;
+    uint32_t num_rows = 0;
+    PUniqueId load_id;
+    load_id.set_hi(0);
+    load_id.set_lo(0);
```
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:64]: unused variables

File: be/src/storage/lake/spark_load.cpp [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    op_write->mutable_rowset()->set_num_rows(writer->num_rows());
+    op_write->mutable_rowset()->set_data_size(writer->data_size());
+    op_write->mutable_rowset()->set_overlapped(op_write->rowset().segments_size() > 1);
+    RETURN_IF_ERROR(cur_tablet.put_txn_log(std::move(txn_log)));
+
+    _write_bytes += static_cast<int64_t>(writer->data_size());
+    _write_rows += static_cast<int64_t>(writer->num_rows());
+    VLOG(10) << "convert delta file end. res=" << st.to_string() << ", tablet=" << cur_tablet.id() << ", processed_rows"
+             << num_rows;
```
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:151]:
```suggestion
    VLOG(10) << "convert delta file end. res=" << st.to_string() << ", tablet=" << cur_tablet.id() << ", processed_rows="
             << writer->num_rows();
```

File: be/src/storage/lake/spark_load.cpp [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+            } else {
+                if (reader->eof()) {
+                    break;
+                }
+
+                ChunkHelper::padding_char_columns(char_field_indexes, schema, *tablet_schema, chunk.get());
+                RETURN_IF_ERROR(writer->write(*chunk));
+                num_rows += chunk->num_rows();
+                chunk->reset();
+            }
```
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:124]:
```suggestion
            if (!st.ok()) {
                LOG(WARNING) << "Fail to read broker: " << st;
                return st;
            } else if (reader->eof()) {
                break;
            } else {
                ChunkHelper::padding_char_columns(char_field_indexes, schema, *tablet_schema, chunk.get());
                RETURN_IF_ERROR(writer->write(*chunk));
                chunk->reset();
            }
```

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 8]
```diff
@@ -117,6 +122,25 @@ AgentStatus EngineBatchLoadTask::_init() {
...
+    auto tablet_manager = ExecEnv::GetInstance()->lake_tablet_manager();
+    auto tablet_or = tablet_manager->get_tablet(_push_req.tablet_id);
+    if (!tablet_or.ok()) {
+        return STARROCKS_PUSH_INVALID_TABLE;
+    }
+
+    DCHECK(!_push_req.__isset.http_file_path);
+    _is_init = true;
+    return status;
+}
```
Comment (sduzh) [inline be/src/storage/task/engine_batch_load_task.cpp:142]: useless function

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 9]
```diff
@@ -151,37 +175,65 @@ Status EngineBatchLoadTask::_push(const TPushReq& request, std::vector<TTabletIn
...
+    Status res;
+
+    if (request.tablet_type == TTabletType::TABLET_TYPE_LAKE) {
+        auto tablet_manager = ExecEnv::GetInstance()->lake_tablet_manager();
+        ASSIGN_OR_RETURN(auto tablet, tablet_manager->get_tablet(request.tablet_id));
+
+        lake::SparkLoadHandler handler;
+        res = handler.process_streaming_ingestion(tablet, request, type, tablet_info_vec);
+        if (!res.ok()) {
+            LOG(WARNING) << "Fail to load file. res=" << res << ", txn_id: " << request.transaction_id
```
Comment (sduzh) [inline be/src/storage/task/engine_batch_load_task.cpp:195]:
```suggestion
            LOG(WARNING) << "Fail to load file. res=" << res << ", txn_id=" << request.transaction_id
```

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 10]
```diff
@@ -151,37 +175,65 @@ Status EngineBatchLoadTask::_push(const TPushReq& request, std::vector<TTabletIn
...
+                         << ", tablet=" << tablet.id()
+                         << ", cost=" << PrettyPrinter::print(duration_ns, TUnit::TIME_NS);
+            StarRocksMetrics::instance()->push_requests_fail_total.increment(1);
+        } else {
+            LOG(INFO) << "Finish to load file."
+                      << ". txn_id: " << request.transaction_id << ", tablet=" << tablet.id()
```
Comment (sduzh) [inline be/src/storage/task/engine_batch_load_task.cpp:201]:
```suggestion
                      << ". txn_id=" << request.transaction_id << ", tablet=" << tablet.id()
```

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 11]
```diff
@@ -151,37 +175,65 @@ Status EngineBatchLoadTask::_push(const TPushReq& request, std::vector<TTabletIn
...
+        if (tablet == nullptr) {
+            LOG(WARNING) << "Not found tablet: " << request.tablet_id;
+            StarRocksMetrics::instance()->push_requests_fail_total.increment(1);
+            return Status::NotFound(fmt::format("Not found tablet: {}", request.tablet_id));
+        }
+
+        vectorized::PushHandler push_handler;
+        res = push_handler.process_streaming_ingestion(tablet, request, type, tablet_info_vec);
+        if (!res.ok()) {
+            LOG(WARNING) << "Fail to load file. res=" << res << ", txn_id: " << request.transaction_id
```
Comment (sduzh) [inline be/src/storage/task/engine_batch_load_task.cpp:222]:
```suggestion
            LOG(WARNING) << "Fail to load file. res=" << res << ", txn_id=" << request.transaction_id
```

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 12]
```diff
@@ -151,37 +175,65 @@ Status EngineBatchLoadTask::_push(const TPushReq& request, std::vector<TTabletIn
...
+                         << ", tablet=" << tablet->full_name()
+                         << ", cost=" << PrettyPrinter::print(duration_ns, TUnit::TIME_NS);
+            StarRocksMetrics::instance()->push_requests_fail_total.increment(1);
+        } else {
+            LOG(INFO) << "Finish to load file."
+                      << ". txn_id: " << request.transaction_id << ", tablet=" << tablet->full_name()
```
Comment (sduzh) [inline be/src/storage/task/engine_batch_load_task.cpp:228]:
```suggestion
                      << ". txn_id=" << request.transaction_id << ", tablet=" << tablet->full_name()
```

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java (hunk 2 above)
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:466]: Remove this log.

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java [hunk 13]
```diff
@@ -521,17 +530,28 @@ private Replica findRelatedReplica(OlapTable olapTable, Partition partition,
...
+            }
+            return allQueryableReplicas.get(0);
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:542]:
```suggestion
            return allQueryableReplicas.isEmpty() ? null : allQueryableReplicas.get(0);
```

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java [hunk 14]
```diff
@@ -521,17 +530,28 @@ private Replica findRelatedReplica(OlapTable olapTable, Partition partition,
...
-                    backendId, tabletId, indexId);
+
+        Tablet tablet = index.getTablet(tabletId);
+        if (tablet instanceof LakeTablet) {
+            List<Replica> allQueryableReplicas = Lists.newArrayList();
+            tablet.getQueryableReplicas(allQueryableReplicas, Collections.emptyList(),
+                    0, -1, 0);
+            if (allQueryableReplicas.isEmpty()) {
+                LOG.warn("no queryable replica found in tablet {}", tabletId);
+            }
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:541]:
```suggestion
            if (allQueryableReplicas.isEmpty()) {
                LOG.warn("no queryable replica found in tablet {}", tabletId);
                return null;
            }
```

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java [hunk 15]
```diff
@@ -475,70 +476,48 @@ private Set<Long> submitPushTasks() throws UserException {
...
-                                for (Replica replica : ((LocalTablet) tablet).getImmutableReplicas()) {
-                                    long replicaId = replica.getId();
-                                    tabletAllReplicas.add(replicaId);
-                                    if (!tabletToSentReplicaPushTask.containsKey(tabletId)
-                                            || !tabletToSentReplicaPushTask.get(tabletId).containsKey(replicaId)) {
+                                PushBrokerReaderParams params = getPushBrokerReaderParams(table, indexId);
+
+                                if (tablet instanceof LocalTablet) {
+                                    // for debug
+                                    LOG.info("tablet is LocalTablet");
```
Comment (sduzh) [inline fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java:486]: Remove this log.

File: be/src/storage/lake/spark_load.cpp [hunk 16]
```diff
@@ -0,0 +1,162 @@
...
+
+    _request = request;
+    Status st = _do_streaming_ingestion(tablet, request, push_type);
+    if (st.ok()) {
+        if (tablet_info_vec != nullptr) {
+            _get_tablet_infos(tablet, tablet_info_vec);
+        }
+        LOG(INFO) << "process realtime vectorized push successfully. "
+                  << "tablet=" << tablet.id() << ", partition_id=" << request.partition_id
+                  << ", txn_id: " << request.transaction_id;
```
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:30]:
```suggestion
                  << ", txn_id=" << request.transaction_id;
```

File: be/src/storage/lake/spark_load.cpp [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+#include "storage/tablet_manager.h"
+#include "storage/lake/filenames.h"
+
+namespace starrocks::lake {
+
+using PushBrokerReader = starrocks::vectorized::PushBrokerReader;
+
+Status SparkLoadHandler::process_streaming_ingestion(const Tablet& tablet, const TPushReq& request, PushType push_type,
+                                                     std::vector<TTabletInfo>* tablet_info_vec) {
+    LOG(INFO) << "begin to realtime vectorized push. tablet=" << tablet.id() << ", txn_id: " << request.transaction_id;
```
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:19]:
```suggestion
    LOG(INFO) << "begin to realtime vectorized push. tablet=" << tablet.id() << ", txn_id=" << request.transaction_id;
```

File: be/src/storage/lake/spark_load.cpp (hunk 4 above)
Comment (sduzh) [inline be/src/storage/lake/spark_load.cpp:56]:
```suggestion
```
if (push_type == PUSH_NORMAL_V2) {
    return _load_convert(tablet, request);
} else {
    return Status::NotSupported(fmt::format("unsupported push type for lake table: {}", push_type));
}
```
```

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 41 / 78 (52.56%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[466, 498, 499, 534, 535, 536, 537, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|40|57|70.18%|[502, 503, 504, 505, 506, 509, 510, 511, 512, 513, 514, 516, 517, 518, 603, 604, 605]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 41 / 77 (53.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[497, 498, 533, 534, 535, 536, 538, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|40|56|71.43%|[499, 500, 501, 502, 505, 506, 507, 508, 509, 510, 512, 513, 514, 599, 600, 601]|

File: be/src/storage/lake/spark_load.cpp [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+    DCHECK(request.__isset.use_vectorized && request.use_vectorized);
+
+    _request = request;
+    Status st = _do_streaming_ingestion(tablet, request, push_type);
+    if (st.ok()) {
+        if (tablet_info_vec != nullptr) {
+            _get_tablet_infos(tablet, tablet_info_vec);
+        }
+        LOG(INFO) << "process realtime vectorized push successfully. "
```
[the remark this answers is not in the document]
Comment (abc982627271) [reply be/src/storage/lake/spark_load.cpp:28]: old code is the same as this

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 41 / 77 (53.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[497, 498, 533, 534, 535, 536, 538, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|40|56|71.43%|[499, 500, 501, 502, 505, 506, 507, 508, 509, 510, 512, 513, 514, 599, 600, 601]|

File: be/src/storage/lake/spark_load.cpp [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+
+    ASSIGN_OR_RETURN(auto tablet_schema, cur_tablet.get_schema());
+    vectorized::Schema schema = ChunkHelper::convert_schema_to_format_v2(*tablet_schema);
+    auto chunk = ChunkHelper::new_chunk(schema, 0);
+    auto char_field_indexes = ChunkHelper::get_char_field_indexes(schema);
+
+    // 2. Init PushBrokerReader to read broker file if exist,
+    //    in case of empty push this will be skipped.
+    std::string path = _request.broker_scan_range.ranges[0].path;
+    LOG(INFO) << "tablet=" << cur_tablet.id() << ", file path=" << path
```
[the remark this answers is not in the document]
Comment (abc982627271) [reply be/src/storage/lake/spark_load.cpp:62]: old logic is the same

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 41 / 77 (53.25%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[497, 498, 533, 534, 535, 536, 538, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|40|56|71.43%|[499, 500, 501, 502, 505, 506, 507, 508, 509, 510, 512, 513, 514, 599, 600, 601]|

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 20]
```diff
@@ -151,37 +154,65 @@ Status EngineBatchLoadTask::_push(const TPushReq& request, std::vector<TTabletIn
...
-        LOG(WARNING) << "Fail to load file. res=" << res << ", txn_id: " << request.transaction_id
-                     << ", tablet=" << tablet->full_name()
-                     << ", cost=" << PrettyPrinter::print(duration_ns, TUnit::TIME_NS);
-        StarRocksMetrics::instance()->push_requests_fail_total.increment(1);
+
+    Status res;
+
+    if (request.tablet_type == TTabletType::TABLET_TYPE_LAKE) {
+        auto tablet_manager = ExecEnv::GetInstance()->lake_tablet_manager();
+        ASSIGN_OR_RETURN(auto tablet, tablet_manager->get_tablet(request.tablet_id));
```
Comment (wyb) [inline be/src/storage/task/engine_batch_load_task.cpp:169]: call `StarRocksMetrics::instance()->push_requests_fail_total.increment(1)` if get tablet failed.

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 21]
```diff
@@ -57,7 +57,10 @@ Status EngineBatchLoadTask::execute() {
 
     AgentStatus status = STARROCKS_SUCCESS;
     if (_push_req.push_type == TPushType::LOAD_V2) {
-        status = _init();
+        if (_push_req.tablet_type != TTabletType::TABLET_TYPE_LAKE) {
+            status = _init();
```
Comment (wyb) [inline be/src/storage/task/engine_batch_load_task.cpp:61]: `_is_init` should be set and will be checked in `_process`.

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java [hunk 22]
```diff
@@ -475,70 +476,44 @@ private Set<Long> submitPushTasks() throws UserException {
...
-                                        tabletFinishedReplicas.add(replicaId);
+                                } else {
+                                    // lake tablet
+                                    List<Replica> allQueryableReplicas = Lists.newArrayList();
+                                    List<Replica> localReplicas = Lists.newArrayList();
+                                    long visibleVersion = partition.getVisibleVersion();
+                                    tablet.getQueryableReplicas(allQueryableReplicas, localReplicas,
+                                            visibleVersion, -1, schemaHash);
+
+                                    for (Replica replica : allQueryableReplicas) {
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java:505]: Only need to send push task to primary backend.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:disappointed: **fail** : 53 / 78 (67.95%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|0|18|00.00%|[497, 498, 533, 534, 535, 536, 538, 539, 540, 542, 544, 545, 546, 548, 549, 550, 551, 553]|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|52|57|91.23%|[499, 516, 591, 592, 593]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 66 / 79 (83.54%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|13|18|72.22%|[497, 498, 542, 545, 546]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|52|58|89.66%|[499, 516, 517, 592, 593, 594]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 66 / 79 (83.54%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|13|18|72.22%|[497, 498, 542, 545, 546]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|52|58|89.66%|[499, 516, 517, 592, 593, 594]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 66 / 79 (83.54%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[86]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|13|18|72.22%|[497, 498, 542, 545, 546]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|52|58|89.66%|[499, 516, 517, 592, 593, 594]|

File: fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java [hunk 23]
```diff
@@ -475,83 +477,49 @@ private Set<Long> submitPushTasks() throws UserException {
...
+                                    long backendId = ((LakeTablet) tablet).getPrimaryBackendId();
+                                    Backend backend = GlobalStateMgr.getCurrentSystemInfo().
+                                            getBackend(backendId);
+                                    if (backend == null) {
+                                        LOG.warn("replica {} not exists", backendId);
+                                        continue;
                                     }
+                                    pushTask(backend.getId(), tableId, partitionId, indexId, tabletId,
+                                            tabletId, schemaHash, params, batchTask, tabletMetaStr,
+                                            backend, new Replica(tableId, backendId, 0, null),
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/load/loadv2/SparkLoadJob.java:521]:
```suggestion
                                            backend, new Replica(tabletId, backendId, 0, null),
```

File: fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java [hunk 24]
```diff
@@ -521,17 +529,29 @@ private Replica findRelatedReplica(OlapTable olapTable, Partition partition,
...
+        Tablet tablet = index.getTablet(tabletId);
+        if (tablet instanceof LakeTablet) {
+            List<Replica> allQueryableReplicas = Lists.newArrayList();
+            tablet.getQueryableReplicas(allQueryableReplicas, Collections.emptyList(),
+                    0, -1, 0);
+            if (allQueryableReplicas.isEmpty()) {
+                LOG.warn("no queryable replica found in tablet {}", tabletId);
+                return null;
+            }
+            return allQueryableReplicas.get(0);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/leader/LeaderImpl.java:542]:
```suggestion
            return new Replica(tablet id, backend id, -1, normal);
```

File: fe/fe-core/src/main/java/com/starrocks/lake/LakeTablet.java [hunk 25]
```diff
@@ -83,7 +83,7 @@ public Set<Long> getBackendIds() {
     public void getQueryableReplicas(List<Replica> allQuerableReplicas, List<Replica> localReplicas,
                                      long visibleVersion, long localBeId, int schemaHash) {
         for (long backendId : getBackendIds()) {
-            Replica replica = new Replica(-1, backendId, -1, null);
+            Replica replica = new Replica(getId(), backendId, -1, null);
```
Comment (wyb) [inline fe/fe-core/src/main/java/com/starrocks/lake/LakeTablet.java:86]:
```suggestion
            Replica replica = new Replica(getId(), backendId, -1, normal);
```

File: be/src/storage/task/engine_batch_load_task.cpp [hunk 26]
```diff
@@ -99,20 +100,24 @@ AgentStatus EngineBatchLoadTask::_init() {
...
+        }
 
-    // check disk capacity
-    if (tablet->data_dir()->capacity_limit_reached(_push_req.__isset.http_file_size)) {
-        return STARROCKS_DISK_REACH_CAPACITY_LIMIT;
+        // check disk capacity
+        if (tablet->data_dir()->capacity_limit_reached(_push_req.__isset.http_file_size)) {
+            return STARROCKS_DISK_REACH_CAPACITY_LIMIT;
+        }
+        DCHECK(!_push_req.__isset.http_file_path);
```
Comment (wyb) [inline be/src/storage/task/engine_batch_load_task.cpp:118]: Both local and lake tablets should check.

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 62 / 74 (83.78%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[88]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|9|13|69.23%|[498, 499, 539, 540]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|52|58|89.66%|[499, 516, 517, 592, 593, 594]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 57 / 69 (82.61%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[88]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|9|13|69.23%|[498, 499, 539, 540]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|47|53|88.68%|[499, 508, 509, 593, 594, 595]|

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 63 / 76 (82.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[88]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|9|13|69.23%|[498, 499, 539, 540]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|53|60|88.33%|[499, 516, 517, 526, 598, 599, 600]|

**Review by wyb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 63 / 76 (82.89%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/lake/LakeTablet.java|0|1|00.00%|[88]|
|:large_blue_circle:|com/starrocks/task/PushTask.java|1|2|50.00%|[183]|
|:large_blue_circle:|com/starrocks/leader/LeaderImpl.java|9|13|69.23%|[498, 499, 539, 540]|
|:large_blue_circle:|com/starrocks/load/loadv2/SparkLoadJob.java|53|60|88.33%|[501, 518, 519, 528, 600, 601, 602]|

**Review by sduzh — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**